File: src/jniglue/tests/org/OpenColorIO/BakerTest.java

import junit.framework.TestCase;
import org.OpenColorIO.*;

public class BakerTest extends TestCase {
    
    String SIMPLE_PROFILE = ""
    + "ocio_profile_version: 1\n"
    + "\n"
    + "strictparsing: false\n"
    + "\n"
    + "colorspaces:\n"
    + "\n"
    + "  - !<ColorSpace>\n"
    + "    name: lnh\n"
    + "    bitdepth: 16f\n"
    + "    isdata: false\n"
    + "    allocation: lg2\n"
    + "\n"
    + "  - !<ColorSpace>\n"
    + "    name: test\n"
    + "    bitdepth: 8ui\n"
    + "    isdata: false\n"
    + "    allocation: uniform\n"
    + "    to_reference: !<ExponentTransform> {value: [2.2, 2.2, 2.2, 1]}\n";
    
    String EXPECTED_LUT = ""
    + "CSPLUTV100\n"
    + "3D\n"
    + "BEGIN METADATA\n"
    + "this is some metadata!\n"
    + "END METADATA\n"
    + "\n"
    + "4\n"
    + "0.000977 0.039373 1.587401 64.000000\n"
    + "0.000000 0.333333 0.666667 1.000000\n"
    + "4\n"
    + "0.000977 0.039373 1.587401 64.000000\n"
    + "0.000000 0.333333 0.666667 1.000000\n"
    + "4\n"
    + "0.000977 0.039373 1.587401 64.000000\n"
    + "0.000000 0.333333 0.666667 1.000000\n"
    + "\n"
    + "2 2 2\n"
    + "0.042823 0.042823 0.042823\n"
    + "6.622026 0.042823 0.042823\n"
    + "0.042823 6.622026 0.042823\n"
    + "6.622026 6.622026 0.042823\n"
    + "0.042823 0.042823 6.622026\n"
    + "6.622026 0.042823 6.622026\n"
    + "0.042823 6.622026 6.622026\n"
    + "6.622026 6.622026 6.622026\n"
    + "\n";
    
    protected void setUp() {
    }
    
    protected void tearDown() {
    }
    
    public void test_interface() {
        Baker bake = new Baker().Create();
        Baker bakee = bake.createEditableCopy();
        Config _cfg = new Config().CreateFromStream(SIMPLE_PROFILE);
        Config cfg = _cfg.createEditableCopy();
        assertEquals(2, cfg.getNumColorSpaces());
        Look foo = new Look().Create();
        foo.setName("foo");
        cfg.addLook(foo);
        Look bar = new Look().Create();
        bar.setName("bar");
        cfg.addLook(bar);
        bakee.setConfig(cfg);
        cfg = bakee.getConfig();
        assertEquals(2, cfg.getNumColorSpaces());
        bakee.setFormat("cinespace");
        assertEquals("cinespace", bakee.getFormat());
        bakee.setType("3D");
        assertEquals("3D", bakee.getType());
        bakee.setMetadata("this is some metadata!");
        assertEquals("this is some metadata!", bakee.getMetadata());
        bakee.setInputSpace("lnh");
        assertEquals("lnh", bakee.getInputSpace());
        bakee.setLooks("foo, -bar");
        assertEquals("foo, -bar", bakee.getLooks());
        bakee.setTargetSpace("test");
        assertEquals("test", bakee.getTargetSpace());
        bakee.setShaperSize(4);
        assertEquals(4, bakee.getShaperSize());
        bakee.setCubeSize(2);
        assertEquals(2, bakee.getCubeSize());
        String output = bakee.bake();
        assertEquals(EXPECTED_LUT, output);
        assertEquals(5, bakee.getNumFormats());
        assertEquals("cinespace", bakee.getFormatNameByIndex(2));
        assertEquals("3dl", bakee.getFormatExtensionByIndex(1));
    }
    
}

File: tests/python/BakerTest.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the OpenColorIO Project.

# TODO: Add getFormatMetadata tests.

import copy, unittest, os, sys
import PyOpenColorIO as OCIO

class BakerTest(unittest.TestCase):

    SIMPLE_PROFILE = """ocio_profile_version: 1

strictparsing: false

displays:
  TestDisplay:
    - !<View> {name: TestView, colorspace: test}

colorspaces:

  - !<ColorSpace>
    name: lnh
    bitdepth: 16f
    isdata: false
    allocation: lg2

  - !<ColorSpace>
    name: test
    bitdepth: 8ui
    isdata: false
    allocation: uniform
    to_reference: !<ExponentTransform> {value: [2.2, 2.2, 2.2, 1]}"""

    EXPECTED_LUT = """CSPLUTV100
3D

BEGIN METADATA
END METADATA

4
0.000977 0.039373 1.587401 64.000000
0.000000 0.333333 0.666667 1.000000
4
0.000977 0.039373 1.587401 64.000000
0.000000 0.333333 0.666667 1.000000
4
0.000977 0.039373 1.587401 64.000000
0.000000 0.333333 0.666667 1.000000

2 2 2
0.042823 0.042823 0.042823
6.622026 0.042823 0.042823
0.042823 6.622026 0.042823
6.622026 6.622026 0.042823
0.042823 0.042823 6.622026
6.622026 0.042823 6.622026
0.042823 6.622026 6.622026
6.622026 6.622026 6.622026

"""

    def assert_lut_match(self, baked, expected):
        lines = baked.splitlines()
        expected_lines = expected.splitlines()
        self.assertEqual(len(lines), len(expected_lines))
        # Text compare for the first lines.
        for i in range(6):
            self.assertEqual(lines[i], expected_lines[i])
        # Compare values after (results might be slightly different on some plaforms).
        for i in range(6, len(lines)):
            # Skip blank lines.
            if lines[i] == '':
                continue
            # Line 16 is the cube size.
            if i == 16:
                self.assertEqual(lines[i], expected_lines[i])
                continue
            lf = lines[i].split(' ')
            elf = expected_lines[i].split(' ')
            for j in range(len(lf)):
                self.assertAlmostEqual(float(lf[j]), float(elf[j]), delta = 0.00001)

    def test_copy(self):
        """
        Test the deepcopy() method.
        """
        cfg = OCIO.Config().CreateFromStream(self.SIMPLE_PROFILE)

        bake = OCIO.Baker()
        bake.setConfig(cfg)
        bake.setFormat("cinespace")
        bake.setInputSpace("lnh")
        bake.setLooks("foo, +bar")
        bake.setTargetSpace("test")
        bake.setShaperSize(4)
        bake.setCubeSize(2)

        other = copy.deepcopy(bake)
        self.assertFalse(other is bake)

        self.assertEqual(other.getConfig(), bake.getConfig())
        self.assertEqual(other.getFormat(), bake.getFormat())
        self.assertEqual(other.getInputSpace(), bake.getInputSpace())
        self.assertEqual(other.getLooks(), bake.getLooks())
        self.assertEqual(other.getTargetSpace(), bake.getTargetSpace())
        self.assertEqual(other.getShaperSize(), bake.getShaperSize())
        self.assertEqual(other.getCubeSize(), bake.getCubeSize())

    def test_interface(self):
        """
        Test similar to C++ CPU test.
        """
        bake = OCIO.Baker()
        cfg = OCIO.Config().CreateFromStream(self.SIMPLE_PROFILE)
        cs = cfg.getColorSpaces()
        self.assertEqual(len(cs), 2)
        bake.setConfig(cfg)
        cfg2 = bake.getConfig()
        cs2 = cfg2.getColorSpaces()
        self.assertEqual(len(cs2), 2)

        # Using target space.
        bake.setFormat("cinespace")
        self.assertEqual("cinespace", bake.getFormat())
        bake.setInputSpace("lnh")
        self.assertEqual("lnh", bake.getInputSpace())
        bake.setLooks("foo, +bar")
        self.assertEqual("foo, +bar", bake.getLooks())
        bake.setLooks("")
        bake.setTargetSpace("test")
        self.assertEqual("test", bake.getTargetSpace())
        bake.setShaperSize(4)
        self.assertEqual(4, bake.getShaperSize())
        bake.setCubeSize(2)
        self.assertEqual(2, bake.getCubeSize())
        output = bake.bake()
        self.assert_lut_match(output, self.EXPECTED_LUT)

        # Using display / view.
        bake = OCIO.Baker()
        bake.setConfig(cfg)
        bake.setFormat("cinespace")
        self.assertEqual("cinespace", bake.getFormat())
        bake.setInputSpace("lnh")
        self.assertEqual("lnh", bake.getInputSpace())
        bake.setDisplayView("TestDisplay", "TestView")
        self.assertEqual("TestDisplay", bake.getDisplay())
        self.assertEqual("TestView", bake.getView())
        bake.setShaperSize(4)
        self.assertEqual(4, bake.getShaperSize())
        bake.setCubeSize(2)
        self.assertEqual(2, bake.getCubeSize())
        output = bake.bake()
        self.assert_lut_match(output, self.EXPECTED_LUT)

        fmts = bake.getFormats()
        self.assertEqual(len(fmts), 12)
        self.assertEqual("cinespace", fmts[4][0])
        self.assertEqual("3dl", fmts[1][1])
